ml/src/main/scala/org/apache/spark/ml/h2o/param/H2OTargetEncoderParams.scala

```diff
@@ -0,0 +1,125 @@
+    require(getInputCols() != null && getInputCols().nonEmpty, "The list of input columns can't be null or empty!")
+    val fields = flatSchema.fields
+    val fieldNames = fields.map(_.name)
+    require(fieldNames.contains(getLabelCol()),
+      s"The specified label column '${getLabelCol()}' was not found in the input dataset!")
+    for(inputCol <- getInputCols()) {
+      require(fieldNames.contains(inputCol),
+        s"The specified input column '$inputCol' was not found in the input dataset!")
+    }
+    val ioIntersection = getInputCols().intersect(getOutputCols())
+    require(ioIntersection.isEmpty,
+      s"""The columns [${ioIntersection.map(i => s"'$i'").mkString(", ")}] are specified
+         |as input columns and also as output columns. There can't be an overlap.""".stripMargin)
+    StructType(flatSchema.fields ++ getOutputCols().map(StructField(_, DoubleType, nullable = true)))
+  }
+
+  //
+  // List of Parameters
+  //
+  protected final val foldCol = new NullableStringParam(this, "foldCol", "Fold column name")
+  protected final val labelCol = new Param[String](this, "labelCol", "Label column name")
+  protected final val inputCols = new StringArrayParam(this, "inputCols", "Names of columns that will be transformed")
+  protected final val holdoutStrategy = new H2OTargetEncoderHoldoutStrategyParam(this,
+    "holdoutStrategy",
+    """A strategy deciding what records will be excluded when calculating the target average on the training dataset.
+      |Options:
+      | None        - All rows are considered for the calculation
+      | LeaveOneOut - All rows except the row the colculation is made for
+      | KMeans      - Only out-of-fold data is considered (The option requires foldCol to be set.)
+    """.stripMargin)
+  protected final val blending = new CaseClassParam[H2OTargetEncoderBlendingSettings](this,
+    "blending",
+    """If set, the target average becomes a weighted average of the group target value and the global target value of a given row.
+      |The weight is determined by the size of the given group that the row belongs to.
+      |Attributes:
+      | InflectionPoint - The bigger number it's, the bigger groups will consider the global target value as a component in the weighted average.
+      | Smoothing       - Controls the rate of transition between group target values and global target values.""".stripMargin)
+  protected final val noise = new CaseClassParam[H2OTargetEncoderNoiseSettings](this,
+    "noise",
+    """The settings affecting how much of noise will be added to the target average.
+      |Attributes:
+      | Amount - amount of random noise
+      | Seed   - a seed of the generator producing the random noise
+    """.stripMargin)
+
+  //
+  // Default values
+  //
+  setDefault(
+    foldCol -> null,
+    labelCol -> "label",
+    inputCols -> Array[String](),
+    holdoutStrategy -> H2OTargetEncoderHoldoutStrategy.None,
+    blending -> null,
```

Comment:
These parameters are serialized by reflection and deserialized by Pyspark. I'm not sure whether there is a support for options already. Maybe @jakubhava could shed some light on that.

Quite like skipping null defaults, but if we do that, we should do the same for all SW algorithms. We need to eventually double check the consequences.

Exposed default values in H2O-3 would be nice, it least it will be always consistent if a H2O-3 default value changes in future.

Reply:
If bleeding have default params then H2O should define them and we should use them. Otherwise use `null` as is in the rest of the code